File: AWS/iac-microbenchmark/lambdas/python/aes256_decrypt.py
import boto3
import base64
import json
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

def lambda_handler(event, context):
    # Initialize KMS client
    kms_client = boto3.client('kms')
    
    body = json.loads(event["body"])

    # Extract the encrypted values from the request body
    encrypted_data_key_b64 = body.get('encrypted_data_key')
    iv_b64 = body.get('iv')
    tag_b64 = body.get('tag')
    encrypted_message_b64 = body.get('encrypted_message')

    # Check if all required values are provided
    if not all([encrypted_data_key_b64, iv_b64, tag_b64, encrypted_message_b64]):
        return {
            'statusCode': 400,
            'headers': {"Access-Control-Allow-Origin": "*", "content-type": "application/json"},
            'body': json.dumps({'error': 'Missing required encrypted data or parameters.'})
        }

    # Decode the base64-encoded values
    encrypted_data_key = base64.b64decode(encrypted_data_key_b64)
    iv = base64.b64decode(iv_b64)
    tag = base64.b64decode(tag_b64)
    ciphertext = base64.b64decode(encrypted_message_b64)
    
    # Decrypt the data key using KMS
    response = kms_client.decrypt(
        CiphertextBlob=encrypted_data_key
    )

    plaintext_data_key = response['Plaintext']
    
    # Initialize decryption using AES-GCM
    cipher = Cipher(algorithms.AES(plaintext_data_key), modes.GCM(iv, tag))
    decryptor = cipher.decryptor()
    
    # Decrypt the message
    plaintext_message = decryptor.update(ciphertext) + decryptor.finalize()
    
    return {
        'statusCode': 200,
        'headers' : {"Access-Control-Allow-Origin": "*",
                     "content-type": "application/json"},
        'body': json.dumps({
            'message': plaintext_message.decode('utf-8')
        })
    }

File: AWS/iac-microbenchmark/lambdas/typescript/aes256_decrypt.ts
import { KMSClient, DecryptCommand } from "@aws-sdk/client-kms";
import { createDecipheriv } from "crypto";
import { Context } from 'aws-lambda';

// handler for AWS Lambda
export const handler = async (event: any, context: Context): Promise<any> => {

  // Initialize the KMS client
  const kmsClient = new KMSClient({ region: 'us-east-1' });

  // Get the KMS-encrypted key, IV, authentication tag, and encrypted message from the event body
  const body = event.body ? JSON.parse(event.body) : {};
  const { encrypted_data_key, iv, tag, encrypted_message } = body;

  if (!encrypted_data_key || !iv || !tag || !encrypted_message) {
    return {
      statusCode: 400,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "content-type": "application/json"
      },
      body: JSON.stringify({ error: "Missing required parameters" }),
    };
  }

  // Decrypt the encrypted data key using KMS
  const decryptCommand = new DecryptCommand({
    CiphertextBlob: Buffer.from(encrypted_data_key, "base64"),
  });

  try {
    const decryptResponse = await kmsClient.send(decryptCommand);
    const plaintextDataKey = decryptResponse.Plaintext as Uint8Array;

    // Convert the IV and tag back from base64
    const ivBuffer = Buffer.from(iv, "base64");
    const tagBuffer = Buffer.from(tag, "base64");
    const encryptedMessageBuffer = Buffer.from(encrypted_message, "base64");

    // Decrypt the message using AES-GCM
    const decipher = createDecipheriv("aes-256-gcm", plaintextDataKey, ivBuffer);
    decipher.setAuthTag(tagBuffer);

    const decryptedMessage = Buffer.concat([
      decipher.update(encryptedMessageBuffer),
      decipher.final(),
    ]).toString("utf-8");

    // Return the decrypted message
    return {
      statusCode: 200,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "content-type": "application/json"
      },
      body: JSON.stringify({
        decrypted_message: decryptedMessage,
      }),
    };

  } catch (error) {
    console.error('Error decrypting message:', error);
    return {
      statusCode: 500,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "content-type": "application/json"
      },
      body: JSON.stringify({ error: 'An error occurred during the decryption process' }),
    };
  }
};
